zkevm-circuits/src/evm_circuit/execution.rs

```diff
@@ -0,0 +1,727 @@
+            rw_counter: usize,
+            is_write: bool,
+        },
+        TxAccessListStorageSlot {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        TxRefund {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        Account {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        AccountStorage {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        AccountDestructed {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        CallContext {
+            rw_counter: usize,
+            is_write: bool,
+        },
+        Stack {
+            rw_counter: usize,
+            is_write: bool,
+            call_id: usize,
+            stack_pointer: usize,
+            value: Word,
+        },
+        Memory {
+            rw_counter: usize,
+            is_write: bool,
+            call_id: usize,
+            memory_address: u64,
+            byte: u8,
+        },
+    }
+
+    impl Rw {
+        pub fn stack_value(&self) -> Word {
+            match self {
+                Self::Stack { value, .. } => *value,
+                _ => unreachable!(),
+            }
+        }
+
+        pub fn table_assignment<F: FieldExt>(&self, randomness: F) -> [F; 8] {
+            match self {
+                Self::Stack {
+                    rw_counter,
+                    is_write,
+                    call_id,
+                    stack_pointer,
+                    value,
+                } => [
+                    F::from(*rw_counter as u64),
+                    F::from(*is_write as u64),
+                    F::from(RwTableTag::Stack as u64),
+                    F::from(*call_id as u64),
+                    F::from(*stack_pointer as u64),
+                    RandomLinearCombination::random_linear_combine(
+                        value.to_le_bytes(),
+                        randomness,
+                    ),
+                    F::zero(),
+                    F::zero(),
+                ],
+                Self::Memory {
+                    rw_counter,
+                    is_write,
+                    call_id,
+                    memory_address,
+                    byte,
+                } => [
+                    F::from(*rw_counter as u64),
+                    F::from(*is_write as u64),
+                    F::from(RwTableTag::Memory as u64),
+                    F::from(*call_id as u64),
+                    F::from(*memory_address),
+                    F::from(*byte as u64),
+                    F::zero(),
+                    F::zero(),
+                ],
+                _ => unimplemented!(),
+            }
+        }
+    }
+}
+use bus_mapping_tmp::{Block, Call, ExecStep, Transaction};
+
+pub(crate) trait ExecutionGadget<F: FieldExt> {
+    const NAME: &'static str;
+
+    const EXECUTION_RESULT: ExecutionResult;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self;
```

Comment:
I have a question here. So all constraints in a `ConstraintBuilder` share the same `condition`. What if we need constraints with different conditions, e.g., `CallDataCopy`? The current trait cannot support such use case.

Reply:
There is a function `condition` in `ConstraintBuilder` for nested branching, I think it's enough to support complicated opcode like `CALLDATACOPY`. Would you like to show some part of the spec that you are concern that current `ConstraintBuilder` cannot support?

Reply:
Ah, I see. I missed the `condition` function. That'd be enough.